File: sdks/python/paycrest_sdk/provider.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional


@dataclass
class ProviderListOrdersQuery:
    """Typed filter object for ``ProviderClient.list_orders``.

    ``from_date`` / ``to_date`` map to the aggregator's ``from`` / ``to``
    query params (named to avoid clashing with Python's ``from`` keyword).
    """

    currency: str
    page: int = 1
    page_size: int = 10
    status: Optional[str] = None
    ordering: Optional[str] = None
    search: Optional[str] = None
    export: Optional[str] = None
    from_date: Optional[str] = None
    to_date: Optional[str] = None


class ProviderClient:
    def __init__(self, http_client):
        self._http = http_client

    def list_orders(
        self,
        query: "Optional[ProviderListOrdersQuery | str]" = None,
        page: int = 1,
        page_size: int = 10,
        status: Optional[str] = None,
        ordering: Optional[str] = None,
        search: Optional[str] = None,
        export: Optional[str] = None,
        from_date: Optional[str] = None,
        to_date: Optional[str] = None,
    ) -> dict:
        """List provider-matched orders.

        The first positional argument accepts either a
        :class:`ProviderListOrdersQuery` value object or a currency code
        string (kept for backwards compatibility with the
        ``list_orders("NGN", page=…)`` signature).
        """
        if isinstance(query, str):
            currency = query
            query = None
        else:
            currency = None

        if query is None:
            if not currency:
                raise ValueError("currency (or ProviderListOrdersQuery) is required")
            query = ProviderListOrdersQuery(
                currency=currency,
                page=page,
                page_size=page_size,
                status=status,
                ordering=ordering,
                search=search,
                export=export,
                from_date=from_date,
                to_date=to_date,
            )
        response = self._http.call(
            "GET",
            "/provider/orders",
            query={
                "currency": query.currency,
                "page": query.page,
                "pageSize": query.page_size,
                "status": query.status,
                "ordering": query.ordering,
                "search": query.search,
                "export": query.export,
                "from": query.from_date,
                "to": query.to_date,
            },
        )
        return response["data"]

    def get_order(self, order_id: str) -> dict:
        response = self._http.call("GET", f"/provider/orders/{order_id}")
        return response["data"]

    def get_stats(self, currency: Optional[str] = None) -> dict:
        response = self._http.call("GET", "/provider/stats", query={"currency": currency})
        return response["data"]

    def get_node_info(self) -> dict:
        response = self._http.call("GET", "/provider/node-info")
        return response["data"]

    def get_market_rate(self, token: str, fiat: str) -> dict:
        response = self._http.call("GET", f"/provider/rates/{token}/{fiat}")
        return response["data"]

File: sdks/go/sdk/provider.go
package sdk

import (
	"context"
	"fmt"
	"net/http"
	"strconv"
)

type ProviderService struct {
	http *httpClientConfig
}

type ProviderListOrdersParams struct {
	Currency string
	Page     int
	PageSize int
	Status   string
	Ordering string
	Search   string
	Export   string
	From     string
	To       string
}

func (p *ProviderService) ListOrders(ctx context.Context, params ProviderListOrdersParams) (*ListOrdersResponse, error) {
	page := params.Page
	if page == 0 {
		page = 1
	}
	pageSize := params.PageSize
	if pageSize == 0 {
		pageSize = 10
	}

	query := map[string]string{
		"currency": params.Currency,
		"page":     strconv.Itoa(page),
		"pageSize": strconv.Itoa(pageSize),
		"status":   params.Status,
		"ordering": params.Ordering,
		"search":   params.Search,
		"export":   params.Export,
		"from":     params.From,
		"to":       params.To,
	}

	var response APIResponse[ListOrdersResponse]
	if err := p.http.request(ctx, http.MethodGet, "/provider/orders", query, nil, &response); err != nil {
		return nil, err
	}
	return &response.Data, nil
}

func (p *ProviderService) GetOrder(ctx context.Context, orderID string) (*PaymentOrder, error) {
	var response APIResponse[PaymentOrder]
	if err := p.http.request(ctx, http.MethodGet, "/provider/orders/"+orderID, nil, nil, &response); err != nil {
		return nil, err
	}
	return &response.Data, nil
}

func (p *ProviderService) GetStats(ctx context.Context, currency string) (*ProviderStats, error) {
	query := map[string]string{"currency": currency}
	var response APIResponse[ProviderStats]
	if err := p.http.request(ctx, http.MethodGet, "/provider/stats", query, nil, &response); err != nil {
		return nil, err
	}
	return &response.Data, nil
}

func (p *ProviderService) GetNodeInfo(ctx context.Context) (map[string]interface{}, error) {
	var response APIResponse[map[string]interface{}]
	if err := p.http.request(ctx, http.MethodGet, "/provider/node-info", nil, nil, &response); err != nil {
		return nil, err
	}
	return response.Data, nil
}

func (p *ProviderService) GetMarketRate(ctx context.Context, token, fiat string) (*MarketRateResponse, error) {
	path := fmt.Sprintf("/provider/rates/%s/%s", token, fiat)
	var response APIResponse[MarketRateResponse]
	if err := p.http.request(ctx, http.MethodGet, path, nil, nil, &response); err != nil {
		return nil, err
	}
	return &response.Data, nil
}
